src/main/java/org/jitsi/videobridge/EndpointMessageTransport.java

```diff
@@ -250,25 +212,21 @@ protected void onSelectedEndpointsChangedEvent(
         Object src,
         JSONObject jsonObject)
     {
-        // Find the new pinned endpoint.
-        Object o = jsonObject.get("selectedEndpoints");
-        if (!(o instanceof JSONArray))
-        {
-            logger.warn("Received invalid or unexpected JSON: " + jsonObject);
-            return;
-        }
+        relayJSONObject(jsonObject);
+    }
 
-        JSONArray jsonArray = (JSONArray) o;
-        Set<String> newSelectedEndpoints = new HashSet<>();
-        for (Object endpointId : jsonArray)
+    private void relayJSONObject(JSONObject jsonObject)
+    {
+        Conference conference = getConference();
+        if (conference == null || conference.isExpired())
         {
-            if (endpointId != null && endpointId instanceof String)
-            {
-                newSelectedEndpoints.add((String)endpointId);
-            }
+            logger.warn(
+                "Unable to relay a JSON object, the conference is null or expired.");
+            return;
         }
 
-        endpoint.selectedEndpointsChanged(newSelectedEndpoints);
+        // Notify Cthulhu and its minions about selected/pinned events.
+        conference.sendMessage(jsonObject.toString(), Collections.EMPTY_LIST, true);
```

Comment:
i'm confused by this: it's in `EndpointMessageTransport`, but only seems to handle the octo case?  shouldn't this be logic in `OctoEndpoindMessageTransport`? or, at least, some other logic to forward to the local endpoints?

Reply:
Thanks Brian, you're right, my code is incomplete; apparently I forgot several super calls super.onXXXEndpointChangedEvent(). I'll update the PR.

I'm still going to add a short explanation here just for completeness and to make sure we agree on this approach and I haven't missed anything.

The code in `EndpointMessageTransport` executes when a message is received by a local endpoint, and so in that case we want to broadcast the message to all the participating bridges.

The code that actually does _stuff_ is moved to `AbstractEndpointMessageTransport` and its shared between local and remote endpoint.

There isn't anything special to do in `OctoEndpoindMessageTransport`.

This PR will in particular fix the `incrementSelectedCount` and `decrementSelectedCount` methods to track inter-bridge counts, so we can try enabling the off-stage participant layer suspension.

Reply:
i just thought that the logic to transmit over octo 'datachannel' vs an endpoint datachannel was different, so was thinking there would still need to be some specific logic in `OctoEndpointMessageTransport`?

Reply:
i had misunderstood how octo messaging works and missed the fact that `AbstractEndpointMessageTransport.endpoint` is null in the case of octo.. d5066f90f14bc399fa356d7bdb13ea0adf8b13c7 adds the octo specific logic.

I was thinking that for message propagation we could wrap the original message in an envelope and add the destination endpoint id as a property of the envelope, instead of adding it as a property to the original message (which is what I do now).